### DISCUSSION TITLE
How to verify if using SHM or multicast

### DISCUSSION THREAD

**Discussion by ipa-rwu** [eProsima/Fast-DDS#4005] [Support]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

I expect there are no messages going to multicast server 239.255.0.1

### Current behavior

I still can see messgae go to 239.255.0.1
```shell
12:46:30.805971 wlp0s20f3 Out IP ThinkPad.40577 > 239.255.0.1.17900: UDP, length 328
	0x0000:  4500 0164 82d4 4000 0111 44a0 c0a8 006c  E..d..@...D....l
	0x0010:  efff 0001 9e81 45ec 0150 b276 5254 5053  ......E..P.vRTPS
	0x0020:  0203 010f 010f 3636 7417 41fc 0100 0000  ......66t.A.....
	0x0030:  0901 0800 6fc6 4c65 d2a9 6d4d 1505 2401  ....o.Le..mM..$.
	0x0040:  0000 1000 0001 00c7 0001 00c2 0000 0000  ................
	0x0050:  0100 0000 0003 0000 1500 0400 0203 0000  ................
	0x0060:  1600 0400 010f 0000 5000 1000 010f 3636  ........P.....66
	0x0070:  7417 41fc 0100 0000 0000 01c1 3200 1800  t.A.........2...
	0x0080:  0100 0000 fa45 0000 0000 0000 0000 0000  .....E..........
	0x0090:  0000 0000 c0a8 006c 3100 1800 0100 0000  .......l1.......
	0x00a0:  fb45 0000 0000 0000 0000 0000 0000 0000  .E..............
	0x00b0:  c0a8 006c 3100 1800 1000 0000 ff45 0000  ...l1........E..
	0x00c0:  5536 3600 0000 0000 0000 0000 0000 0000  U66.............
	0x00d0:  0200 0800 1400 0000 0000 0000 5800 0400  ............X...
	0x00e0:  3f0c 3f0c 6200 2000 1a00 0000 2f74 616c  ?.?.b......./tal
	0x00f0:  6b65 725f 6c69 7374 656e 6572 2f6c 6973  ker_listener/lis
	0x0100:  7465 6e65 7200 0000 2c00 2800 2300 0000  tener...,.(.#...
	0x0110:  656e 636c 6176 653d 2f74 616c 6b65 725f  enclave=/talker_
	0x0120:  6c69 7374 656e 6572 2f6c 6973 7465 6e65  listener/listene
	0x0130:  723b 0000 5900 2800 0100 0000 1100 0000  r;..Y.(.........
	0x0140:  5041 5254 4943 4950 414e 545f 5459 5045  PARTICIPANT_TYPE
	0x0150:  0000 0000 0700 0000 5349 4d50 4c45 0000  ........SIMPLE..
	0x0160:  0100 0000                                ....
12:46:33.759891 lo    In  IP localhost.54962 > 239.255.0.1.17900: UDP, length 324
	0x0000:  4500 0160 67d3 4000 0111 a1b8 7f00 0001  E..`g.@.........
	0x0010:  efff 0001 d6b2 45ec 014c 705f 5254 5053  ......E..Lp_RTPS
	0x0020:  0203 010f 010f 3636 6a17 70fa 0100 0000  ......66j.p.....
	0x0030:  0901 0800 6fc6 4c65 be97 a841 1505 2001  ....o.Le...A....
	0x0040:  0000 1000 0001 00c7 0001 00c2 0000 0000  ................
	0x0050:  0100 0000 0003 0000 1500 0400 0203 0000  ................
	0x0060:  1600 0400 010f 0000 5000 1000 010f 3636  ........P.....66
	0x0070:  6a17 70fa 0100 0000 0000 01c1 3200 1800  j.p.........2...
	0x0080:  0100 0000 f845 0000 0000 0000 0000 0000  .....E..........
	0x0090:  0000 0000 c0a8 006c 3100 1800 0100 0000  .......l1.......
	0x00a0:  f945 0000 0000 0000 0000 0000 0000 0000  .E..............
	0x00b0:  c0a8 006c 3100 1800 1000 0000 fd45 0000  ...l1........E..
	0x00c0:  5536 3600 0000 0000 0000 0000 0000 0000  U66.............
	0x00d0:  0200 0800 1400 0000 0000 0000 5800 0400  ............X...
	0x00e0:  3f0c 3f0c 6200 1c00 1800 0000 2f74 616c  ?.?.b......./tal
	0x00f0:  6b65 725f 6c69 7374 656e 6572 2f74 616c  ker_listener/tal
	0x0100:  6b65 7200 2c00 2800 2100 0000 656e 636c  ker.,.(.!...encl
	0x0110:  6176 653d 2f74 616c 6b65 725f 6c69 7374  ave=/talker_list
	0x0120:  656e 6572 2f74 616c 6b65 723b 0000 0000  ener/talker;....
	0x0130:  5900 2800 0100 0000 1100 0000 5041 5254  Y.(.........PART
	0x0140:  4943 4950 414e 545f 5459 5045 0000 0000  ICIPANT_TYPE....
	0x0150:  0700 0000 5349 4d50 4c45 0000 0100 0000  ....SIMPLE......
```
![Screenshot from 2023-11-09 13-02-37](https://github.com/eProsima/Fast-DDS/assets/54799760/4dec70c3-eefc-4b06-80dd-8abb4c12af86)
![image](https://github.com/eProsima/Fast-DDS/assets/54799760/20531cfb-71bd-4eab-aedf-53c093cc3082)

### Steps to reproduce

I created a [test](https://github.com/ipa-rwu/ros2_dds_container_test ) here.
You can use [compose.shm.yaml](https://github.com/ipa-rwu/ros2_dds_container_test/blob/main/compose.shm.yaml) to start.

I run a talker and a listener in different docker containers. I want to use **Shared Memory Transport** 
Based on the information here [Leveraging Fast DDS SHM in Docker deployments](https://fast-dds.docs.eprosima.com/en/v2.12.0/docker/shm_docker.html)
I use
```yaml
    ipc: host
    network_mode: host
    pid: host
```
and I also mount "/dev/shm" which I think it is not necessary since I set "ipc: host".

But I still also can see traffic going through 239.255.0.1.

I also try to disable multicast by giving wrong locator ip address following [Disabling multicast discovery ](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/wifi/disable_multicast.html).
You can find dds configuration here: [shm_config_wrong_peer.xml](https://github.com/ipa-rwu/ros2_dds_container_test/blob/main/shm_config_wrong_peer.xml)

I expect listener still can get messges because of using SHM but it doesn't. 

I also try out fastdds monitor. I can see talker and listener have identical 4 first bytes on guid as you can see my screenshort.

How can I know if it really uses SHM? 

### Fast DDS version/commit

```cmd
ros2 doctor --report | grep fastrtps
```
results:
```shell
fastrtps_cmake_module                     : latest=2.2.1, local=2.2.1
rmw_fastrtps_shared_cpp                   : latest=6.2.4, local=6.2.4
rosidl_typesupport_fastrtps_cpp           : latest=2.2.1, local=2.2.1
rosidl_typesupport_fastrtps_c             : latest=2.2.1, local=2.2.1
rmw_fastrtps_cpp                          : latest=6.2.4, local=6.2.4
middleware name    : rmw_fastrtps_cpp
```

### Platform/Architecture

Other. Please specify in Additional context section.

### Transport layer

Default configuration, UDPv4 & SHM

### Additional context

I am using Ubuntu 22.04.3 amd64

### XML configuration file

```xml
<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
    <transport_descriptors>
        <!-- Create a descriptor for the new transport -->
        <transport_descriptor>
            <transport_id>shm_transport</transport_id>
            <type>SHM</type>
        </transport_descriptor>
    </transport_descriptors>

    <participant profile_name="SHMParticipant" is_default_profile="true">
        <rtps>
            <!-- Link the Transport Layer to the Participant -->
            <userTransports>
                <transport_id>shm_transport</transport_id>
            </userTransports>

        </rtps>
    </participant>
</profiles>

<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
    <transport_descriptors>
        <!-- Create a descriptor for the new transport -->
        <transport_descriptor>
            <transport_id>shm_transport</transport_id>
            <type>SHM</type>
        </transport_descriptor>
    </transport_descriptors>

    <participant profile_name="SHMParticipant" is_default_profile="true">
        <rtps>
            <!-- Link the Transport Layer to the Participant -->
            <userTransports>
                <transport_id>shm_transport</transport_id>
            </userTransports>
            <builtin>
                <!-- Choosing a specific unicast address -->
                <metatrafficUnicastLocatorList>
                    <locator>
                        <udpv4>
                            <address>192.168.10.13</address>
                            <port>7412</port>
                        </udpv4>
                    </locator>
                </metatrafficUnicastLocatorList>
            </builtin>
        </rtps>
    </participant>
</profiles>
```

### Network traffic capture

![image](https://github.com/eProsima/Fast-DDS/assets/54799760/ddcd30ec-d695-412f-9dcc-c9aa3106e226)

**Comment by JesusPoderoso** [comment]
Hi @ipa-rwu, thanks for using Fast DDS. 
In the XML configuration file that you have shared, you are enabling the SHM transport, but the default Fast DDS behavior makes your application to perform the discovery phase using UDP. 
If you want to force all communication (discovery traffic plus user data traffic) to be through SHM, you need to disable the use of builtin transports.
That can be done in the XML configuration as follows:
```
<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
    <transport_descriptors>
        <!-- Create a descriptor for the new transport -->
        <transport_descriptor>
            <transport_id>shm_transport</transport_id>
            <type>SHM</type>
        </transport_descriptor>
    </transport_descriptors>

    <participant profile_name="SHMParticipant" is_default_profile="true">
        <rtps>
            <!-- Link the Transport Layer to the Participant -->
            <userTransports>
                <transport_id>shm_transport</transport_id>
            </userTransports>
            <useBuiltinTransports>false</useBuiltinTransports>
        </rtps>
    </participant>
</profiles>
```

**Comment by ipa-rwu** [comment]
@JesusPoderoso Thank you for your help. Is there a way to monitoring SHM? For cyclonedds I can use (iceoryx introspection)[https://iceoryx.io/v1.0.1/getting-started/examples/icecrystal/] to monitor shared memory segments. 
I only find `fastdds shm clean` to clean up

**Comment by JesusPoderoso** [comment]
Yes @ipa-rwu, you can configure the [`rtps_dump_file_`](https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/shared_memory/shared_memory.html#sharedmemtransportdescriptor). If it is not empty, all the Shared Memory traffic is traced to a file. The output file format is _tcpdump_ hexadecimal text, and can be processed with protocol analyzer applications such as Wireshark. Specifically, to open the file using **Wireshark**, use the “Import from Hex Dump” option using the “Raw IPv4” encapsulation type.

**Reply by ipa-rwu** [reply]
Thank you very much. It works for me.

**Comment by JesusPoderoso** [comment]
I am moving this question to the support discussion forum as per [Fast DDS CONTRIBUTING guidelines](https://github.com/eProsima/Fast-DDS/blob/master/CONTRIBUTING.md).
